CMakeLists.txt

```diff
@@ -41,7 +41,7 @@ set(CMAKE_C_STANDARD_REQUIRED ON)
 # may be useful to simplify the build and distribution of the application. E.g.
 # in test farms, or in cases where the end-user system configs are known in
 # advance. Use it with discretion.
-option (BUILD_SHARED_LIBS "Build shared libs" ON)
+option (OPENCL_ICD_LOADER_BUILD_SHARED_LIBS "Build shared libs" ON)
```

Comment:
I've played around with `BUILD_SHARED_LIBS` in a subproject scenario and indeed they don't mix well (at least when both the calling project the the called project wish to control it. Whomever gets there first (because it's a `CACHE` variable) wins (without `FORCE` that is).

As far as the grand scheme of things go the OpenCL-ICD-Loader can only guarantee build script API compatibility with itself. If a project involves building another into their own, they take on following all the changes the project makes. (It's simply impossible for a project to be backward compatible with _every_ project recursing into.) Having that said...

**For OpenCL-ICD-Loader to control this global property manually is a bug.** If a calling project favors static libs, and before us calls `option(BUILD_SHARED_LIBS "Don't touch" OFF)` the ICD-Loader by default will be built as a static library. The ICD Loader's preference to favor dynamic libraries is _indifferent to what our caller prefers or asked for_. Therefor breaking those who were in the shoes of @LeonBrands, who had to correct global state after recursing into our folder is not a break but a fix to a latent bug.

Suggestion is the following:

```suggestion
include(CMakeDependentOption)
if(DEFINED BUILD_SHARED_LIBS)
  set(OPENCL_ICD_LOADER_BUILD_SHARED_LIBS_DEFAULT ${BUILD_SHARED_LIBS})
else()
  set(OPENCL_ICD_LOADER_BUILD_SHARED_LIBS_DEFAULT ON)
endif()
  option(OPENCL_ICD_LOADER_BUILD_SHARED_LIBS "Build OpenCL ICD Loader as shared library" ${OPENCL_ICD_LOADER_BUILD_SHARED_LIBS_DEFAULT})
```

This way:
- People who consumed the ICD Loader as a separate install
  - If they don't ask for anything, they get the same `SHARED` as they used to.
  - If they ask for `SHARED` or `STATIC` like they used to using `BUILD_SHARED_LIBS`, they get the lib type they expect.
- People who consume the ICD by making the build part of theirs
  - If they set `BUILD_SHARED_LIBS` to `ON` or `OFF`, the loader is built as `SHARED`/`STATIC` as the global setting implies.
  - If they don't set anything, _BUGFIX_ the new option when leaking won't alter the global default (which is building `STATIC` when `BUILD_SHARED_LIBS` isn't set to anything).
  - If they wish to mix and match _NEW FEATURE_ only then and only they have to opt in via the new variable `OPENCL_ICD_LOADER_BUILD_SHARED_LIBS`.

This way people get the escape hatch they want but existing practice will not change.

_(ps.: the reason why this code is ugly is because unlike most variables, `BUILD_SHARED_LIBS` has no default value, the default isn't `OFF`, but it's absence implies `OFF` too, therefor the need for `OPENCL_ICD_LOADER_BUILD_SHARED_LIBS_DEFAULT`.)_

Reply:
Thank you for your response! This is a good solution that preserves the global state and OpenCL's preference for dynamic loaders 👍